gordon_janitor_gcp/gce_clients.py

```diff
@@ -0,0 +1,297 @@
+                             project,
+                             page_size=500,
+                             instance_filter=None,
+                             retries=3,
+                             retry_wait=5):
+        """Fetch all instances in a GCE project.
+
+        Endpoint documentation:
+        https://cloud.google.com/compute/docs/reference/latest/instances/aggregatedList
+
+        Args:
+            project (str): unique, user-provided project ID.
+            page_size (int): hint for the client to only retrieve up to this
+                number of results per API call.
+            instance_filter (str): endpoint-specific filter string used to
+                retrieve a subset of instances. This is passed directly to the
+                endpoint's "filter" URL query parameter.
+        Returns:
+            list: dicts containing instance data.
+        """
+        url = f'{self.BASE_URL}/{project}/aggregated/instances'
+        params = {'maxResults': page_size}
+        if instance_filter:
+            params['filter'] = instance_filter
+
+        return await self.list_all_items(url, params, retries, retry_wait)
+
+    def parse_response_into_items(self, response, items):
+        """Parse API response for all instances in a project into a list.
+
+        Args:
+            response (dict): endpoint response data.
+            items (list): list for collecting parsed results.
+        """
+        zones = response.get('items', {})
+        for zone in zones.values():
+            for instance in zone.get('instances', []):
+                if not any([
+                        self._blacklisted_by_tag(instance),
+                        self._blacklisted_by_metadata(instance)
+                ]):
+                    try:
+                        items.append(self._extract_instance_data(instance))
+                    except KeyError as e:
+                        logging.error(
+                            'Could not extract instance information for '
+                            f'{instance} because of missing key {e}, skipping.')
+
+    def _extract_instance_data(self, instance):
+        return {
+            'hostname': instance['name'],
+            'internal_ip': instance['networkInterfaces'][0]['networkIP'],
+            'external_ip': instance['networkInterfaces'][0]['accessConfigs'][0]['natIP'],  # noqa: E501
+        }
+
+    def _blacklisted_by_tag(self, instance):
+        """Check if instance has any blacklisted tags.
+
+        Args:
+            instance (dict): instance data as returned by API.
+        Return:
+            bool: if instance has any blacklisted tag.
+        """
+        instance_tags = instance['tags'].get('items', [])
+        for tag in instance_tags:
+            if tag in self.blacklisted_tags:
+                msg = (f'Instance "{instance["name"]}" filtered out for '
+                       f'blacklisted tag: "{tag}"')
+                logging.error(msg)
```

Comment:
Any reason this can't be `info`? I don't even think it constitutes as a `warning`.